src/problems/mod.rs

```diff
@@ -1,100 +1,382 @@
-pub trait HasKnownOptimum {
+/// impl Problem for Sphere {
+///     type Encoding = Vec<f64>;
+///     type Objective = SingleObjective;
+///
+///     fn name(&self) -> &str {
+///         "Sphere"
+///     }
+/// }
+///
+/// impl KnownOptimumProblem for Sphere {
+///     const DELTA: f64 = 1e-10;
+///
+///     fn known_optimum(&self) -> SingleObjective {
+///         0.0.try_into().unwrap()
+///     }
+/// }
+///
+/// // A value greater than 1e-10 does not counts as optimum.
+/// let sphere = Sphere { dim: 1 };
+/// assert!(!sphere.optimum_reached(1e-9.try_into().unwrap()));
+/// ```
+pub trait KnownOptimumProblem: SingleObjectiveProblem {
+    /// A constant representing the tolerance level for comparing objective values.
+    ///
+    /// This value is used as tolerance to automatically implement [`OptimumReachedProblem`].
+    const DELTA: f64 = 1e-8;
+
+    /// Retrieves the known optimum objective value for the optimization problem.
     fn known_optimum(&self) -> SingleObjective;
 }
+
+impl<P: KnownOptimumProblem> OptimumReachedProblem for P {
+    fn optimum_reached(&self, objective: SingleObjective) -> bool {
+        let provided = objective.value();
+        let known = self.known_optimum().value();
+        debug_assert!(
+            provided >= known,
+            "the provided objective value is smaller than the known optimum"
+        );
+
+        (provided - known).abs() <= P::DELTA
+    }
+}
+
+/// The Travelling Salesperson Problem (TSP) optimization problem.
+///
+/// TSP is a single-objective optimization problem that involves finding
+/// the shortest possible route that visits a given set of vertices and
+/// returns to the starting vertex.
+///
+/// # Examples
+///
+/// A simple implementation of the Travelling Salesperson Problem given a edge weight matrix:
+///
+/// ```
+/// use mahf::{Problem, problems::{VectorProblem, TravellingSalespersonProblem}, SingleObjective};
+///
+/// pub struct TSP {
+///     pub instance: String,
+///     pub edge_weights: Vec<Vec<f64>>,
+/// }
+///
+/// impl Problem for TSP {
+///     type Encoding = Vec<usize>;
+///     type Objective = SingleObjective;
+///
+///     fn name(&self) -> &str {
+///         &self.instance
+///     }
+/// }
+///
+/// impl VectorProblem for TSP {
+///     type Element = usize;
+///
+///     fn dimension(&self) -> usize {
+///         self.edge_weights.len()
+///     }
+/// }
+///
+/// impl TravellingSalespersonProblem for TSP {
+///     fn distance(&self, edge: (usize, usize)) -> f64 {
+///         let (source, target) = edge;
+///         self.edge_weights[source][target]
+///     }
+/// }
+/// ```
+pub trait TravellingSalespersonProblem:
+    SingleObjectiveProblem + VectorProblem<Element = usize>
+{
+    /// Calculates the distance between two locations,
+    /// i.e. the edge weight between two vertices identified by their indices.
+    fn distance(&self, edge: (usize, usize)) -> f64;
+}
```

Comment:
This probably should not exist? It's fine for now, but maybe open an issue for its removal.

Reply:
Well, the ACO components need to access the distance information somehow, so I don't see what's wrong with it. Otherwise, we couldn't implement them here. Also, I think it's nice to decouple things somewhat, I much prefer a `P: TravellingSalespersonProblem` bound on TSP specific components over them being specific to e.g. ``SymmetricTSP`.

Reply:
But how does ACO work with other problems? I haven't done any research regarding this, but I think it's being used for other tasks as well, right?

Reply:
Classical ACO (i.e. ACO for combinatorial problems) actually needs all optimization problems in graph form with distance information between the nodes.

There are continuous versions, but I have not looked into them in detail.

Reply:
ACO was initially developed for TSP (which is what we implement) and the components extracted from it are inherently specific to the graph representation of TSP (e.g. `NxN` pheromone matrix for `N` cities). The concept of ACO was of course transferred to other problems and representations, but this also results in a (completely) different implementation (e.g. versions of binary ACO have a `2xN` pheromone matrix for a binary string of length `N`).

Reply:
Huh, if ACO is this specific, we can keep the trait I guess 👍🏻